### PR TITLE
Enable support for image class in the runtime.

### PR DESCRIPTION
**Pull request by alexeyvoronov**
- SYCLMemObjT(interface class for SYCL memory objects) was renamed to SYCLMemObjI.
- The general functionality and fields of the image and buffer classes were moved into new parent class SYCLMemObjT.
- Extend the plugin interface to support image manipulation.
- Enabled image interoperability constructor, set_final_data and set_write_back image class methods.
- Added the handling of user-provided pointer, and allocating aligned memory in image class constructors.
- Added support for image class to work with copy and fill handler class methods.
- Added support for image class in the memory manager.
- Make buffer/image class constructor from a shared pointer to follow "4.7.2.3 Buffer Synchronization Rules" from SYCL specification.

### CODE REVIEW

**Comment by alexeyvoronov** [comment]
Re-based.

**Comment by romanovvlad** [comment]
Update. Add "Mem" to PI APIs that operates with buffer and images.

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+    Requirement Req(Offset, AccessRange, MemoryRange, AccessMode, SYCLMemObject,
+                    Dims, ElemSize);
+    Req.MData = Ptr;
+
+    EventImplPtr Event = Scheduler::getInstance().addCopyBack(&Req);
+    return Event;
+  }
+
+  void finish() {
```
Comment (bader) [inline sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp]: "Finish" SYCL memory object.
What does it mean?

File: sycl/include/CL/sycl/detail/buffer_impl.hpp [hunk 2]
```diff
@@ -340,76 +146,20 @@ template <typename AllocatorT> class buffer_impl : public SYCLMemObjI {
...
-  cl_mem MInteropMemObject = nullptr;
-
-  void *MUserPtr = nullptr;
-  void *MShadowCopy = nullptr;
-  size_t MSizeInBytes = 0;
-
-  property_list MProps;
-  std::function<void(void)> MUploadDataFn = nullptr;
-  AllocatorT MAllocator;
+  ~buffer_impl() { BaseT::finish(); }
```
Comment (bader) [inline sycl/include/CL/sycl/detail/buffer_impl.hpp]: Why we call "finish" method instead of `~BaseT()`?

File: sycl/include/CL/sycl/detail/common.hpp [hunk 3]
```diff
@@ -132,10 +132,6 @@ template <template <int> class T> struct InitializedVal<3, T> {
   template <int Val> static T<3> &&get() { return T<3>{Val, Val, Val}; }
 };
 
-// Fills the lack of enable_if_t in C++11.
-template <bool Cond, typename T = void>
-using enable_if_t = typename std::enable_if<Cond, T>::type;
-
```
Comment (bader) [inline sycl/include/CL/sycl/detail/common.hpp]: This change is not related to the commit, please, move refactoring adding type_traits file to a separate commit.

File: sycl/include/CL/sycl/property_list.hpp [hunk 4]
```diff
@@ -8,14 +8,14 @@
 
 #pragma once
 
+#include <CL/sycl/context.hpp>
```
Comment (bader) [inline sycl/include/CL/sycl/property_list.hpp]: Why do we add this include here?
It's the only change in this file. It's unclear what problem we are solving with this change.

File: sycl/source/detail/image_impl.cpp [hunk 5]
```diff
@@ -76,6 +76,158 @@ uint8_t getImageElementSize(uint8_t NumChannels, image_channel_type Type) {
...
+    return image_channel_order::intensity;
+  case PI_LUMINANCE:
+    return image_channel_order::luminance;
+  case PI_ABGR:
+    return image_channel_order::abgr;
+  default: {
+    assert(!"Unhandled image_channel_order");
+    return static_cast<image_channel_order>(0);
+  }
+  }
```
Comment (bader) [inline sycl/source/detail/image_impl.cpp]: I think conversion from OpenCL values to PI values should be done by utilities defined in OpenCL plug-in rather than here.

File: sycl/include/CL/sycl/handler.hpp [hunk 6]
```diff
@@ -895,16 +895,32 @@ class handler {
   }
 
   // Explicit copy operations API
+  template <access::target AccessTarget>
+  constexpr static bool isConstOrGlobal() {
```
Comment (bader) [inline sycl/include/CL/sycl/handler.hpp]: I thought that one of the advantages of constexpr functions is that we can avoid templates.
Please, make `AccessTarget` a function parameter rather than template parameter.

File: sycl/test/basic_tests/image_api.cpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+// RUNx: env SYCL_DEVICE_TYPE=HOST %t2.out
+// RUN: env SYCL_DEVICE_TYPE=HOST %t3.out
+// RUN: %CPU_RUN_PLACEHOLDER %t1.out
+// RUNx: %CPU_RUN_PLACEHOLDER %t2.out
+// RUN: %GPU_RUN_PLACEHOLDER %t1.out
+// RUNx: %GPU_RUN_PLACEHOLDER %t2.out
+// RUN: %ACC_RUN_PLACEHOLDER %t1.out
+// RUNx: %ACC_RUN_PLACEHOLDER %t2.out
+
+// TODO Turn on t2.out test cases when SYCL image class api will be implemented
```
Comment (bader) [inline sycl/test/basic_tests/image_api.cpp]: I suggest remove them from this patch and add them with "image class api implementation."

File: sycl/test/basic_tests/image_api.cpp [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+                                   Range);
+
+    s::queue Queue;
+    s::device Device = Queue.get_device();
+    if(!Device.get_info<s::info::device::image_support>()) {
+      std::cout << "Images are not supported. The result can not be checked."
+                << std::endl;
+      return 0;
+    }
+#ifdef SYCL_API
```
Comment (bader) [inline sycl/test/basic_tests/image_api.cpp]: Where is this macro defined?
Looks like a dead code to me.
Suggest removing.

**Comment by bader** [comment]
Add [NFC] tag to "[SYCL] Rename releaseMemBuf method to releaseMemObj." commit.

File: sycl/source/detail/image_impl.cpp (hunk 5 above)
Comment (romanovvlad) [reply sycl/source/detail/image_impl.cpp]: These functions convert  from sycl values to PI ones.

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+    EventImplPtr Event = Scheduler::getInstance().addCopyBack(&Req);
+    return Event;
+  }
+
+  // Update host with the latest data + notify scheduler that the memory object
+  // is going to die. After this method is finished no further operations with
+  // the memory object is allowed. This method is executed from child's
+  // destructor. This cannot be done in SYCLMemObjT's destructor as child's
+  // members must be alive.
+  void writeBack() {
```
Comment (bader) [inline sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp]:
```suggestion
  void updateHostMemory() {
```

Thanks for the comment.

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+  // Function which update host with final data on memory object destruction.
+  std::function<void(void)> MUploadDataFunctor;
+  // Field which holds user's shared_ptr in case of memory object is created
+  // using constructor with shared_ptr.
+  shared_ptr_class<const void> MSharedPtrStorage;
+};
+
+} // namespace detail
+} // namespace sycl
+} // namespace cl
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp]: Missing end-of-line...

File: sycl/include/CL/sycl/detail/type_traits.hpp [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+};
+
+template <typename T>
+struct is_output_iterator<T, output_iterator_requirements<T>> {
+  static constexpr bool value = true;
+};
+
+} // namespace detail
+} // namespace sycl
+} // namespace cl
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/type_traits.hpp]: Missing end-of-line...

File: sycl/include/CL/sycl/detail/pi.h [hunk 12]
```diff
@@ -88,6 +88,65 @@ typedef enum {
...
+  PI_MEM_OBJECT_BUFFER         = CL_MEM_OBJECT_BUFFER,
+  PI_MEM_OBJECT_IMAGE2D        = CL_MEM_OBJECT_IMAGE2D,
+  PI_MEM_OBJECT_IMAGE3D        = CL_MEM_OBJECT_IMAGE3D,
+  PI_MEM_OBJECT_IMAGE2D_ARRAY  = CL_MEM_OBJECT_IMAGE2D_ARRAY,
+  PI_MEM_OBJECT_IMAGE1D        = CL_MEM_OBJECT_IMAGE1D,
+  PI_MEM_OBJECT_IMAGE1D_ARRAY  = CL_MEM_OBJECT_IMAGE1D_ARRAY,
+  PI_MEM_OBJECT_IMAGE1D_BUFFER = CL_MEM_OBJECT_IMAGE1D_BUFFER
+} _pi_mem_object_type;
+
+// TODO: populate
```
Comment (Ruyk) [inline sycl/include/CL/sycl/detail/pi.h]: could this "populate" be more specific? What is missing so other people can contribute to it?

File: sycl/include/CL/sycl/detail/pi.h [hunk 13]
```diff
@@ -321,23 +405,40 @@ pi_result piQueueFinish(pi_queue command_queue);
...
 
+pi_mem piMemImageCreate( // TODO: change interface to return error code
+  pi_context              context,
+  pi_mem_flags            flags,
+  const pi_image_format * image_format,
+  const pi_image_desc *   image_desc,
+  void *                  host_ptr,
+  pi_result *             errcode_ret);
+
 pi_result piMemGetInfo(
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h]: Given you are adding piMemImageGetInfo below, I assume this interface is not OK for images, and then it should be changed to piMemBufferGetInfo

File: sycl/include/CL/sycl/detail/pi.h [hunk 14]
```diff
@@ -321,23 +405,40 @@ pi_result piQueueFinish(pi_queue command_queue);
...
-pi_result piMemRetain(pi_mem);
+pi_result piMemImageGetInfo (
+  pi_mem          image,
+  pi_image_info   param_name,
+  size_t          param_value_size,
+  void *          param_value ,
+  size_t *        param_value_size_ret);
 
-pi_result piMemRelease(pi_mem);
+pi_result piMemRetain(
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h]: Is this (and piMemRelease below) suitable for both buffers and images?

**Comment by bader** [comment]
@romanovvlad, please, resolve comments and source code conflicts.

File: sycl/include/CL/sycl/detail/accessor_impl.hpp [hunk 15]
```diff
@@ -10,7 +10,7 @@
 
 #include <CL/sycl/access/access.hpp>
 #include <CL/sycl/detail/event_impl.hpp>
-#include <CL/sycl/detail/sycl_mem_obj.hpp>
+#include <CL/sycl/detail/sycl_mem_obj_i.hpp>
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/accessor_impl.hpp]: Interface.

File: sycl/include/CL/sycl/detail/pi.h (hunk 12 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/pi.h]: Removed TODOs as it's populated enough to support things SYCL needs.

File: sycl/include/CL/sycl/detail/pi.h (hunk 13 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/pi.h]: piMemImageGetInfo works for both images and buffers and provides common info. piMemImageGetInfo provides image specific info.

File: sycl/include/CL/sycl/detail/pi.h (hunk 14 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/pi.h]: That's correct.

**Comment by romanovvlad** [comment]
> 
> 
> Just some minor comments.
> Is host image support also added by this patch ? e.g, host accessor to an image?
Just partially, there is not support for read/write operation yet.

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp (hunk 10 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp]: Not sure how to fix it, I don't see a problem. Could you clarify?

File: sycl/include/CL/sycl/detail/image_impl.hpp [hunk 16]
```diff
@@ -239,12 +223,40 @@ class image_impl : public SYCLMemObjT {
...
     setPitches(Pitch);
-    handleHostData(HData);
+    BaseT::handleHostData(HData, detail::getNextPowerOfTwo(MElementSize));
+  }
+
+  image_impl(cl_mem MemObject, const context &SyclContext,
+             const size_t SizeInBytes, event AvailableEvent = {})
+      : BaseT(MemObject, SyclContext, std::move(AvailableEvent)) {
+    RT::PiMem Mem = pi::pi_cast<RT::PiMem>(BaseT::MInteropMemObject);
+    PI_CALL(RT::piMemGetInfo(Mem, CL_MEM_SIZE, sizeof(size_t), &BaseT::MSizeInBytes,
```
Comment (againull) [inline sycl/include/CL/sycl/detail/image_impl.hpp]: &BaseT::MSizeInBytes is a pointer-to-member creation syntax and it is not what is expected here. Please change to &(BaseT::MSizeInBytes)

File: sycl/include/CL/sycl/detail/image_impl.hpp [hunk 17]
```diff
@@ -239,12 +223,40 @@ class image_impl : public SYCLMemObjT {
...
+      : BaseT(PropList, Allocator), MRange(ImageRange), MOrder(Order),
+        MType(Type), MNumChannels(getImageNumberChannels(MOrder)),
         MElementSize(getImageElementSize(MNumChannels, MType)) {
     setPitches(Pitch);
-    handleHostData(HData);
+    BaseT::handleHostData(HData, detail::getNextPowerOfTwo(MElementSize));
+  }
+
+  image_impl(cl_mem MemObject, const context &SyclContext,
+             const size_t SizeInBytes, event AvailableEvent = {})
```
Comment (againull) [inline sycl/include/CL/sycl/detail/image_impl.hpp]: Please remove unused "const size_t SizeInBytes" parameter.

File: sycl/include/CL/sycl/detail/pi.h (hunk 14 above)
Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h]: OK

File: sycl/include/CL/sycl/detail/pi.h (hunk 13 above)
Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h]: OK

File: sycl/include/CL/sycl/detail/pi.h [hunk 18]
```diff
@@ -88,6 +88,61 @@ typedef enum {
   PI_QUEUE_INFO_REFERENCE_COUNT = CL_QUEUE_REFERENCE_COUNT
 } _pi_queue_info;
 
+typedef enum {
+  PI_IMAGE_FORMAT       = CL_IMAGE_FORMAT,
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h]: please follow the convention that constants name repeat it's type name (e.g. see PI_QUEUE_INFO_DEVICE above). this would be PI_IMAGE_INFO_FORMAT

File: sycl/include/CL/sycl/detail/pi.h [hunk 19]
```diff
@@ -88,6 +88,61 @@ typedef enum {
...
+
+typedef enum {
+  PI_MEM_OBJECT_BUFFER         = CL_MEM_OBJECT_BUFFER,
+  PI_MEM_OBJECT_IMAGE2D        = CL_MEM_OBJECT_IMAGE2D,
+  PI_MEM_OBJECT_IMAGE3D        = CL_MEM_OBJECT_IMAGE3D,
+  PI_MEM_OBJECT_IMAGE2D_ARRAY  = CL_MEM_OBJECT_IMAGE2D_ARRAY,
+  PI_MEM_OBJECT_IMAGE1D        = CL_MEM_OBJECT_IMAGE1D,
+  PI_MEM_OBJECT_IMAGE1D_ARRAY  = CL_MEM_OBJECT_IMAGE1D_ARRAY,
+  PI_MEM_OBJECT_IMAGE1D_BUFFER = CL_MEM_OBJECT_IMAGE1D_BUFFER
+} _pi_mem_object_type;
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h]: I suggest pi_mem_type for shortness

File: sycl/include/CL/sycl/detail/pi.h [hunk 20]
```diff
@@ -88,6 +88,61 @@ typedef enum {
...
+  PI_RGBA      = CL_RGBA,
+  PI_BGRA      = CL_BGRA,
+  PI_ARGB      = CL_ARGB,
+  PI_ABGR      = CL_ABGR,
+  PI_INTENSITY = CL_INTENSITY,
+  PI_LUMINANCE = CL_LUMINANCE,
+  PI_Rx        = CL_Rx,
+  PI_RGx       = CL_RGx,
+  PI_RGBx      = CL_RGBx
+} _pi_channel_order;
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h]: Is this for image only? let's name _pi_image_channel_order

File: sycl/include/CL/sycl/detail/pi.h [hunk 21]
```diff
@@ -88,6 +88,61 @@ typedef enum {
...
+  PI_UNORM_INT_101010 = CL_UNORM_INT_101010,
+  PI_SIGNED_INT8      = CL_SIGNED_INT8,
+  PI_SIGNED_INT16     = CL_SIGNED_INT16,
+  PI_SIGNED_INT32     = CL_SIGNED_INT32,
+  PI_UNSIGNED_INT8    = CL_UNSIGNED_INT8,
+  PI_UNSIGNED_INT16   = CL_UNSIGNED_INT16,
+  PI_UNSIGNED_INT32   = CL_UNSIGNED_INT32,
+  PI_HALF_FLOAT       = CL_HALF_FLOAT,
+  PI_FLOAT            = CL_FLOAT
+} _pi_channel_type;
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h]: _pi_image_channel_type

File: sycl/include/CL/sycl/detail/pi.hpp [hunk 22]
```diff
@@ -43,6 +43,12 @@ namespace pi {
   using PiMemFlags            = ::pi_mem_flags;
   using PiEvent               = ::pi_event;
   using PiSampler             = ::pi_sampler;
+  using PiImageFormat         = ::pi_image_format;
+  using PiImageDesc           = ::pi_image_desc;
+  using PiImageInfo           = ::pi_image_info;
+  using PiMemObjectType       = ::pi_mem_object_type;
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.hpp]: PiImageInfo & PiMemObjectType are unused? Can we remove these?
At least I would like to get read of all object in memobject

File: sycl/include/CL/sycl/detail/image_impl.hpp [hunk 23]
```diff
@@ -309,69 +327,187 @@ class image_impl : public SYCLMemObjT {
...
+  template <typename T, typename... Args>
+  bool checkAny(const T Value, Args... Arguments) {
+    return checkAnyImpl(Value, Arguments...);
   }
 
-  // This utility api is currently used by accessor to get the element size of
-  // the image. Element size is dependent on num of channels and channel type.
-  // This information is not accessible from the image using any public API.
-  uint8_t getElementSize() const { return MElementSize; };
+  RT::PiMemObjectType getClImageType() {
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/image_impl.hpp]: get rid of Cl in the method names

File: sycl/include/CL/sycl/detail/pi.hpp (hunk 22 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/pi.hpp]: They are used in ./include/CL/sycl/detail/image_impl.hpp

File: sycl/include/CL/sycl/detail/pi.h [hunk 24]
```diff
@@ -106,7 +106,7 @@ typedef enum {
   PI_MEM_OBJECT_IMAGE1D        = CL_MEM_OBJECT_IMAGE1D,
   PI_MEM_OBJECT_IMAGE1D_ARRAY  = CL_MEM_OBJECT_IMAGE1D_ARRAY,
   PI_MEM_OBJECT_IMAGE1D_BUFFER = CL_MEM_OBJECT_IMAGE1D_BUFFER
-} _pi_mem_object_type;
+} _pi_mem_type;
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h]: rename PI_MEM_OBJECT_... to PI_MEM_TYPE...

File: sycl/include/CL/sycl/detail/pi.h [hunk 25]
```diff
@@ -106,7 +106,7 @@ typedef enum {
...
 
 typedef enum {
   PI_A         = CL_A,
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h]: PI_A -> PI_IMAGE_CHANNEL_ORDER_A (and others)

File: sycl/include/CL/sycl/detail/pi.h [hunk 26]
```diff
@@ -123,7 +123,7 @@ typedef enum {
   PI_Rx        = CL_Rx,
   PI_RGx       = CL_RGx,
   PI_RGBx      = CL_RGBx
-} _pi_channel_order;
+} _pi_image_channel_order;
 
 typedef enum {
   PI_SNORM_INT8       = CL_SNORM_INT8,
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h]: PI_SNORM_INT8  -> PI_IMAGE_CHANNEL_TYPE_SNORM_INT8

File: sycl/include/CL/sycl/detail/pi.h (hunk 24 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/pi.h]: Can we resolve it in a separate patch?

Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h]: OK, no problem. But please follow up asap, before people get a chance to use it.

**Comment by alexeyvoronov** [comment]
**The order of the commits was broken. Between some commits, SYCL will not compile.**

Why do we need a separate commit "Apply comments related to renaming"? Let's do this in the corresponding commits.
